plutus-tx-plugin/src/Language/PlutusTx/Plugin.hs

```diff
@@ -82,43 +122,32 @@ unconditionally which we pretty much are.
 See https://gitlab.haskell.org/ghc/ghc/issues/16615 for upstream discussion.
 -}
 
-install :: [GHC.CommandLineOption] -> [GHC.CoreToDo] -> GHC.CoreM [GHC.CoreToDo]
-install args todo = do
-    flags <- GHC.getDynFlags
-    let opts = PluginOptions {
-            poDoTypecheck = notElem "dont-typecheck" args
-            , poDeferErrors = elem "defer-errors" args
-            , poContextLevel = if elem "no-context" args then 0 else if elem "debug-context" args then 3 else 1
-            , poDumpPir = elem "dump-pir" args
-            , poDumpPlc = elem "dump-plc" args
-            , poOptimize = notElem "dont-optimize" args
-            }
-        pass = GHC.CoreDoPluginPass "Core to PLC" (pluginPass opts)
-        -- See Note [Making sure unfoldings are present]
-        mode = GHC.SimplMode {
-                    GHC.sm_names = ["Ensure unfoldings are present"]
-                  , GHC.sm_phase = GHC.InitialPhase
-                  , GHC.sm_dflags = flags
-                  , GHC.sm_rules = False
-                  -- You might think you would need this, but apparently not
-                  , GHC.sm_inline = False
-                  , GHC.sm_case_case = False
-                  , GHC.sm_eta_expand = False
-                  }
-        simpl = GHC.CoreDoSimplify 1 mode
-    pure $ simpl:pass:todo
-
-pluginPass :: PluginOptions -> GHC.ModGuts -> GHC.CoreM GHC.ModGuts
-pluginPass opts guts = do
+-- Instantiates our plugin given the plugin options and the found marker symbol
+mkPluginPass :: PluginOptions -> Maybe GHC.Name -> GHC.CoreToDo
+mkPluginPass opts = \case
+    Just markerName -> GHC.CoreDoPluginPass "Core to PLC" $
+                  let ?markerName = markerName
+                      ?opts = opts
+                  in compileMod
+    Nothing -> GHC.CoreDoNothing -- nothing to do
+
+-- | The plugin works at haskell-module level granularity; the plugin
+-- looks at the module's the top-level bindings for plc markers and compile their inner core-expressions.
+compileMod :: (?markerName :: GHC.Name, ?opts :: PluginOptions)
+           => GHC.ModGuts -> GHC.CoreM GHC.ModGuts
+compileMod guts = do
     -- Family env code borrowed from SimplCore
     p_fam_env <- GHC.getPackageFamInstEnv
-    let fam_envs = (p_fam_env, GHC.mg_fam_inst_env guts)
+    let ?famEnvs = (p_fam_env, GHC.mg_fam_inst_env guts)
+      -- start looking for plc calls from the top-level binds
+      in GHC.bindsOnlyPass (runECore . mapM compileBind) guts
 
-    maybeName <- getMarkerName
-    case maybeName of
-        -- nothing to do
-        Nothing   -> pure guts
-        Just name -> GHC.bindsOnlyPass (mapM $ compileMarkedExprsBind (opts, fam_envs) name) guts
+-- | The monad where the plugin runs in for each module.
+-- It is a core->core compiler monad, called CoreM, augmented with pure errors.
+type ECore uni = ExceptT (CompileError uni) GHC.CoreM
```

Comment:
The usual convention is to call a concrete monad something M. The current one I guess is supposed to suggest it handles errors, but this is actually a bit restrictive since if e.g. we added a `ReaderT` it would become misleading. How about `PluginM`?